src/nubank/auth/mod.rs

```diff
@@ -0,0 +1,182 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{self, Read};
+
+use reqwest;
+use serde::Deserialize;
+use serde_json::{self, json, Value};
+
+mod custom_request_builder;
+use custom_request_builder::*;
+
+#[derive(Debug)]
+pub enum AuthError {
+    DiscoveryRequestFailed(reqwest::Error),
+    DiscoveryRequestDecodingFailed(reqwest::Error),
+    DiscoveryJsonConversionFailed(serde_json::Error),
+    UrlNameNotFoundInDiscoveryJson(String),
+    CannotReadCertificate(io::Error),
+    CannotInterpretCertificate(reqwest::Error),
+    CannotBuildClient(reqwest::Error),
+    AuthRequestFailed(reqwest::Error),
+    AuthRequestDecodingFailed(reqwest::Error),
+    AuthJsonConversionFailed(serde_json::Error),
+    RequiredFieldsNotFoundInAuthJson(AuthDataDTO),
+}
+
+#[derive(Deserialize, Debug, Clone)]
+struct Href {
+    href: String,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct AuthDataDTO {
+    access_token: String,
+    #[serde(rename = "_links")]
+    links: HashMap<String, Href>,
+}
+
+#[derive(Debug, Clone)]
+pub struct AuthData {
+    access_token: String,
+    feed_url: String,
+    bills_url: String,
+    customer_url: String,
+    query_url: String,
+    revoke_token_url: String,
+}
+
+pub async fn authenticate(
+    path: String,
+    cpf: String,
+    password: String,
+) -> Result<AuthData, AuthError> {
+    let url = get_url("token".to_string()).await?;
+
+    let id = get_identity(path)?;
+    let client = build_client(id)?;
+
+    let payload = build_payload(cpf, password);
+
+    let response = make_auth_request(client, url, payload).await?;
+    let auth_data_dto = read_request_output(response)?;
+    let auth_data = build_auth_data_obj(auth_data_dto)?;
+
+    Ok(auth_data)
+}
+
+async fn get_url(name: String) -> Result<String, AuthError> {
+    let discovery_app_url = "https://prod-s0-webapp-proxy.nubank.com.br/api/app/discovery";
+
+    let urls_str = reqwest::get(discovery_app_url)
+        .await
+        .map_err(|x| AuthError::DiscoveryRequestFailed(x))?
+        .text()
+        .await
+        .map_err(|x| AuthError::DiscoveryRequestDecodingFailed(x))?;
+
+    let urls = serde_json::from_str::<Value>(&*urls_str)
+        .map_err(|x| AuthError::DiscoveryJsonConversionFailed(x))?;
+
+    match urls.get(&name) {
+        Some(v) => Ok(v.to_string().replace("\"", "")),
+        _ => return Err(AuthError::UrlNameNotFoundInDiscoveryJson(name)),
+    }
+}
+
+fn get_identity(path: String) -> Result<reqwest::Identity, AuthError> {
+    let mut buf = Vec::new();
+    let _ = File::open(path)
+        .map_err(|x| AuthError::CannotReadCertificate(x))?
+        .read_to_end(&mut buf);
+    let id = reqwest::Identity::from_pkcs12_der(&buf, "")
+        .map_err(|x| AuthError::CannotInterpretCertificate(x))?;
+    Ok(id)
```

Comment:
You don't need to use error propagation here, just return the result

Reply:
Done.